execution-engine/src/hcall/common.rs

```diff
@@ -275,578 +160,57 @@ impl<T> From<std::sync::PoisonError<T>> for HostProvisioningError {
     }
 }
 
+impl From<wasmi::Error> for HostProvisioningError {
+    fn from(error: wasmi::Error) -> Self {
+        HostProvisioningError::WasmiError(format!("{:?}", error))
+    }
+}
+
 ////////////////////////////////////////////////////////////////////////////////
 // The Veracruz provisioning state.
 ////////////////////////////////////////////////////////////////////////////////
 
-/// The state of the Veracruz machine, which captures metadata as the Veracruz
-/// state is gradually "provisioned" by the data and program providers.  Also
-/// contains enough data to properly implement the Veracruz H-calls.
+/// A wrapper for VFS, which provides common API used by execution engine.
 #[derive(Clone)]
-pub struct HostProvisioningState<Module, Memory> {
-    /// The data sources that have been provisioned into the machine.
-    data_sources: Vec<DataSourceMetadata>,
-    /// The stream sources that have been provisioned into the machine.
-    stream_sources: Vec<DataSourceMetadata>,
-    /// The expected list of data sources, derived from the global policy
-    /// parameterising the computation.
-    expected_data_sources: Vec<u64>,
-    /// The expected list of stream sources, derived from the global policy
-    /// parameterising the computation.
-    expected_stream_sources: Vec<u64>,
-    /// The current lifecycle state of the machine.
-    lifecycle_state: LifecycleState,
-    /// A reference to the WASM program module that will actually execute on
-    /// the input data sources.
-    program_module: Option<Module>,
-    /// The SHA-256 digest of the bytes of the loaded program, if any.
-    program_digest: Option<Vec<u8>>,
-    /// A reference to the WASM program's linear memory (or "heap").
-    memory: Option<Memory>,
-    /// The result of the previous execution.
-    previous_result: Option<Vec<u8>>,
-    /// The result of the WASM program's computation on the input sources above.
-    result: Option<Vec<u8>>,
-    /// The list of clients (their IDs) that can request shutdown of the
-    /// Veracruz platform.
-    expected_shutdown_sources: Vec<u64>,
+pub struct VFSService {
```

Comment:
Why is this `Arc<Mutex<...>>` field needed?

Reply:
Yeah the ldea is Veracruz runtime instance will hold a VFS Service. Most API in runtime for clients, e.g. provisioning data and program, directly use VFS Service. Only when the execution is called, VFS will be load into the execution engine, by when there are two instances hold the VFSService,